dask/utils.py

```diff
@@ -135,15 +141,29 @@ def filetexts(d, open=open):
             os.remove(filename)
 
 
-opens = {'gzip': gzip.open}
+compressions = {'gz': 'gzip', 'bz2': 'bz2'}
+
+
+def infer_compression(filename):
+    extension = os.path.splitext(filename)[-1].strip('.')
+    return compressions.get(extension, None)
+
+
+opens = {'gzip': GzipFile, 'bz2': BZ2File}
 
 
-def get_bom(fn):
+def open(filename, mode='rb', compression=None, **kwargs):
```

Comment:
Is there a way to make ``open`` compatible with ``with``-statement contexts while still being able to use it as a function? Like the builtin ``open`` function. The problem is that ``BZ2File`` and ``GzipFile`` do not implement the ``__exit__`` method. Otherwise, it would work with the function as it is now. This would just be syntactic sugar though.